cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -287,18 +277,188 @@ void ProcessListScalar(const ScalarType& scalar, ValidFunc&& valid_func,
     int64_t list_length = scalar_data->length;
     const uint8_t* scalar_validity_buf = scalar_data->buffers[0]->data() == NULLPTR ?
       NULLPTR : scalar_data->buffers[0]->data();
-    const uint8_t* scalar_data_buf = scalar_data->buffers[1]->data();
+    const uint8_t* scalar_data_buf_1 = scalar_data->buffers[1]->data() == NULLPTR ?
+      NULLPTR: scalar_data->buffers[1]->data();
+    const uint8_t* scalar_data_buf_2 =
+      scalar_data->buffers.size() < 3 || scalar_data->buffers[2]->data() == NULLPTR ?
+      NULLPTR : scalar_data->buffers[2]->data();
 
-    valid_func(list_length, scalar_validity_buf, scalar_data_buf);
+    valid_func(list_length, scalar_validity_buf, scalar_data_buf_1, scalar_data_buf_2);
   } else { // data.scalar->is_valid == false
     null_func();
   }
 }
 
+// A template to know how much more to allocate for the encoding of a list type for a
+// batch.
+// get_list_length must take
+// - the starting offset of the list
+// - the ending offset of the list
+// - the start of the validity buffer of the nested-type
+// - the start of the nested-type's buffers[1]
+// - the start of the nested-type's buffers[2] or NULLPTR if it does not exist
+// - returns: the amount of bytes to allocate for the list; get_list_length(0, 0, NULLPTR,
+//   NULLPTR, NULLPTR) must return the amount of space to allocate for a null list
+template<typename T, typename LengthFunc>
+void AddListLength(const ExecValue& data, int64_t batch_length, int32_t* lengths,
+                  LengthFunc&& get_list_length) {
+  using Offset = typename T::offset_type;
+  using ScalarType = typename TypeTraits<T>::ScalarType;
+
+  const int32_t null_list_length = get_list_length(0, 0, NULLPTR, NULLPTR, NULLPTR);
+  if (data.is_array()) {
+    const ArraySpan& arr = data.array;
+    const uint8_t* child_validity_buffer = arr.child_data[0].GetValues<uint8_t>(0);
+    const uint8_t* child_data_buffer_1 = arr.child_data[0].GetValues<uint8_t>(1);
+    const uint8_t* child_data_buffer_2 = arr.child_data[0].GetValues<uint8_t>(2);
```

Comment:
I read the code and I do not need to check if it is a NULLPTR as `GetValues` does that already.

Reply:
Also, I had this idea of putting these buffer arrays into a helper function like `ProcessListArray`, but did not do it because it felt like over abstracting and reduces readability.

Reply:
I think that sounds reasonable. 

The point of my earlier suggestions was to separate out the common list-encoding logic from the fixed width list/var length list specific logic to avoid duplicating the former. Pulling out the buffers from the `ExecValue` seems simple enough that we can skip abstracting it. 

If you did want to do this, I think it would make sense to have a helper function on the encoder class that takes in an `ExecValue` and returns the child data buffer for the fixed-width encoder, or the child offset and child data buffers for the var-length encoders. In principle, `AddListLengths` and `EncodeList` shouldn't have to look at the child buffers at all.